### ISSUE TITLE
contextual embeddings of individual words

### ISSUE THREAD

**Issue by Jinsong-Chen** [AnswerDotAI/ModernBERT#191]
I need contextual embeddings of individual words for analysis. However, it turns out that ModernBERT performs much worse than BERT. I use the right version of transformer mentioned in Hugging Face and the following code (same for BERT which works well) to obtain the word-level embeddings. Could there be any problem?

`
from transformers import AutoTokenizer, AutoModel
llm = 'answerdotai/ModernBERT-large'
tokenizer = AutoTokenizer.from_pretrained(llm)
model = AutoModel.from_pretrained(llm)

def embedding(text, tokenizer, model):
    input = tokenizer(texts, padding="longest", truncation=True, max_length=512, return_tensors="pt")
    with torch.no_grad():
        outputs = model(**input)
    return outputs.last_hidden_state

pos_adj=1
text = f"{textA}{word}{textB}'{doc}'"
start_pos = len(tokenizer.tokenize(textA)) + pos_adj
end_pos = len(tokenizer.tokenize(textA + word)) + pos_adj
all_embeddings = embedding(text, tokenizer, model)
word_embedding = torch.mean(all_embeddings[0, start_pos:end_pos], dim=0)
`

**Comment by Jinsong-Chen** [comment]
I am not sure if this is related, but the embedding dimension is only 1k instead of 8k:

'''
>>> word_embedding.shape
torch.Size([1024])
'''

How can I obtain the 8k dimension of embedding?

**Comment by mattlam** [comment]
Might be related to this? https://github.com/AnswerDotAI/ModernBERT/issues/149

**Comment by NohTow** [comment]
Yeah, I don't know what you are evaluating exactly, but if you have _per word_ metrics, it might be related to the issue mentioned.

We should have fixed the tokenizer in the latest version of transformers, could you try installing it (maybe from source just to be sure) and to use `is_split_into_words=True` while calling the tokenizer and also maybe use `add_prefix_space=True` when creating the tokenizer.

Could you try checking the tokenization with BERT and ModernBERT to make sure they are similar?

**Comment by Jinsong-Chen** [comment]
> Yeah, I don't know what you are evaluating exactly, but if you have _per word_ metrics, it might be related to the issue mentioned.
> 
> We should have fixed the tokenizer in the latest version of transformers, could you try installing it (maybe from source just to be sure) and to use `is_split_into_words=True` while calling the tokenizer and also maybe use `add_prefix_space=True` when creating the tokenizer.
> 
> Could you try checking the tokenization with BERT and ModernBERT to make sure they are similar?

We've tried all methods you mentioned with the latest version of transformer, and made sure the tokenization are similar, but ModernBERT still perform worse than BERT. I am not sure if this issue is related to #149 mentioned by mattlam, though.

Specifically, we want to compute the cosine similarity (or dot product) between a word given some context and the context. Ideally, the similarity should be high for a word given relevant context and low otherwise. Note that the word might not appear in the context. Here is a simple case:
words = ['zinc', 'byte']
contexts = ['This metallic element, atomic number 30, plays a vital role in human health and industry. Found in the Earth’s crust, it’s extracted from minerals like sphalerite and used to galvanize steel, preventing corrosion. In biology, it supports immune function, wound healing, and DNA synthesis, with deficiencies leading to growth issues or weakened immunity. Foods like oysters, beef, and pumpkin seeds are rich sources. Industrially, it’s alloyed with copper to form brass, valued for its durability and luster. Discovered in ancient times, this versatile substance remains essential, bridging metallurgy and medicine in surprising ways—all without rusting into obscurity.',
'In the digital realm, information flows as tiny units, each a cluster of eight bits. This fundamental building block, born from computing’s early days, holds a single character—like a letter or number—within its binary embrace. Coined in 1956 by Werner Buchholz, it powers everything from simple text to complex software, with a single one ranging from 0 to 255 in value. Millions or billions of these units combine to form files, programs, and networks, driving modern technology. Though small, their collective might shapes our world, encoding data in a language machines whisper tirelessly, unseen yet ever-present.']

We can use something like "find the contextual meaning of '{word}' given the context '{context}'" to rephrase a new context containing the word, and then obtain the contextual embedding of the word using the above code. The embedding of the context will be obtained independently (e.g., using mean pooling).

Using BERT, we get:
similarity(words, contexts)=array([[0.51822186, 0.40917629],
       [0.42299721, 0.53309619]])

Using ModernBERT, we get:
array([[0.72226554, 0.66950274],
       [0.71057755, 0.59522253]])

For better performance, the difference between the diagonal (relevant cases) and off-diagonal  (irrelevant cases) should be larger (and positive).
Is there any way to improve ModernBERT's performance?

**Comment by NohTow** [comment]
> We can use something like "find the contextual meaning of '{word}' given the context '{context}'"
I am not sure this is totally true, encoders for now are not really promptable/instructable, especially base models.
I have high doubts that the base models can be used successfully for this task without proper training/leveraging another type of trained models.
Maybe you should try using models fine-tuned for retrieval/clustering, such as [ModernBERT-embed](https://huggingface.co/nomic-ai/modernbert-embed-base) and twist a bit your approach to leverage those properties, but this make the comparison harder between BERT and ModernBERT.

But please keep in mind that those models are base models and are not really meant to be used off-the-shelf without specific training.

**Comment by Jinsong-Chen** [comment]
> > We can use something like "find the contextual meaning of '{word}' given the context '{context}'"
> > I am not sure this is totally true, encoders for now are not really promptable/instructable, especially base models.
> > I have high doubts that the base models can be used successfully for this task without proper training/leveraging another type of trained models.
> > Maybe you should try using models fine-tuned for retrieval/clustering, such as [ModernBERT-embed](https://huggingface.co/nomic-ai/modernbert-embed-base) and twist a bit your approach to leverage those properties, but this make the comparison harder between BERT and ModernBERT.
> 
> But please keep in mind that those models are base models and are not really meant to be used off-the-shelf without specific training.

Thanks and we'll try to fine-tune or retrain ModernBERT later. One interesting finding was that BERT large (without any fine-tuning) is the best model so far, among all text embedding models, even including those in the MTEB leaderboard (https://huggingface.co/spaces/mteb/leaderboard). This is quite surprising and my wild guess is that BERT (large)'s embedding is more sensitive to contextual difference than any other model (note that we only need encoder and no prompt or instruction is needed, per se).

I thought that the raw ModernBERT (large) should be similar to, if not better than, the raw BERT (large) anyway. Could there be anything wrong?
